src/Nethermind/Nethermind.Consensus/Stateless/StatelessBlockProcessingEnv.cs

```diff
@@ -47,8 +51,11 @@ public IWorldState WorldState
 
     private IBlockProcessor GetProcessor()
     {
-        IBlockTree statelessBlockTree = new StatelessBlockTree(witness.DecodedHeaders);
-        ITransactionProcessor txProcessor = CreateTransactionProcessor(WorldState, statelessBlockTree);
+        Dictionary<Hash256, BlockHeader> headersByHash = witness.DecodedHeaders.ToDictionary(header => header.Hash ?? throw new ArgumentNullException(), header => header);
+        Dictionary<long, BlockHeader> headersByNumber = witness.DecodedHeaders.ToDictionary(header => header.Number, header => header);
+
+        IBlockTree statelessBlockTree = new StatelessBlockTree(headersByHash, headersByNumber);
```

Comment:
I don't like that we create these dictionaries outside `StatelessBlockTree`. `StatelessBlockTree` user can put whatever `headersByHash` and `headersByNumber` he wants. This is why I suggested to put two interfaces into one class. I think we can implement `IBlockhashCache` for `StatelessBlockTree`.